opm/simulators/linalg/bda/amgclSolverBackend.cpp

```diff
@@ -0,0 +1,332 @@
+                    A_vals[idx] = vals[ij*bs*bs + r*bs + c];
+                    idx++;
+                }
+            }
+        }
+    }
+
+    if (verbosity >= 3) {
+        std::ostringstream out;
+        out << "amgclSolverBackend::convert_data(): " << t.stop() << " s";
+        OpmLog::info(out.str());
+    }
+} // end convert_data()
+
+#if HAVE_VEXCL
+template <typename vexcl_matrix_type, typename vexcl_vector_type, unsigned int block_size>
+void solve_vexcl(
+    const auto A,
+    const boost::property_tree::ptree prm,
+    const vex::Context& ctx,
+    double *b,
+    std::vector<double>& x,
+    const int N,
+    int& iters,
+    double& error)
+{
+    typedef amgcl::backend::vexcl<vexcl_matrix_type> Backend;
+    typedef amgcl::make_solver<amgcl::runtime::preconditioner<Backend>, amgcl::runtime::solver::wrapper<Backend> > Solver;
+
+    typename Solver::backend_params bprm;
+    bprm.q = ctx;  // set vexcl context
+
+    Solver solve(A, prm, bprm); // create solver
+
+    auto b_ptr = reinterpret_cast<vexcl_vector_type*>(b);
+    auto x_ptr = reinterpret_cast<vexcl_vector_type*>(x.data());
+    vex::vector<vexcl_vector_type> B(ctx, N / block_size, b_ptr);
+    vex::vector<vexcl_vector_type> X(ctx, N / block_size, x_ptr);
+
+    std::tie(iters, error) = solve(B, X); // actually perform solve
+
+    vex::copy(X, x_ptr);
+}
+#endif
+
+template <unsigned int block_size>
+void amgclSolverBackend<block_size>::solve_system(double *b, BdaResult &res) {
+    Timer t;
+
+    try {
+        if (backend_type == Amgcl_backend_type::cuda) { // use CUDA
+#if HAVE_CUDA
+            solve_cuda(b);
+#endif
+        } else if (backend_type == Amgcl_backend_type::cpu) { // use builtin backend (CPU)
+            // create matrix object
+            auto Atmp = std::tie(N, A_rows, A_cols, A_vals);
+            auto A = amgcl::adapter::block_matrix<dmat_type>(Atmp);
+
+            // create solver and construct preconditioner
+            // don't reuse this unless the preconditioner can be reused
+            CPU_Solver solve(A, prm);
+
+            // print solver structure (once)
+            std::call_once(print_info, [&](){
+                std::ostringstream out;
+                out << solve << std::endl;
+                OpmLog::info(out.str());
+            });
+
+            // reset x vector
+            std::fill(x.begin(), x.end(), 0.0);
+
+            // create blocked vectors
+            auto b_ptr = reinterpret_cast<dvec_type*>(b);
+            auto x_ptr = reinterpret_cast<dvec_type*>(x.data());
+            auto B = amgcl::make_iterator_range(b_ptr, b_ptr + N / block_size);
+            auto X = amgcl::make_iterator_range(x_ptr, x_ptr + N / block_size);
+
+            // actually solve
+            std::tie(iters, error) = solve(B, X);
+        } else if (backend_type == Amgcl_backend_type::vexcl) {
+#if HAVE_VEXCL
+            vex::Context ctx(vex::Filter::Env && vex::Filter::Count(1));
```

Comment:
The device filter should probably take `--opencl-device-id` into account. See https://vexcl.readthedocs.io/en/latest/initialize.html#device-filters for possible filters. Also, `Filter::Env` here allows controlling the device choice with environment variables.

The context initialization should probably be done once, as it is done for cuda in `cuda_solve`.